scico/solver.py

```diff
@@ -229,25 +230,36 @@ def minimize(
         min_func = _wrap_func(func_, x0_shape, x0_dtype)
         jac = False
 
-    res = spopt.minimize(
-        min_func,
-        x0=x0,
-        args=args,
-        jac=jac,
-        method=method,
-        options=options,
+    res_dict = {}
+
+    def fun(_):
+        min_res = spopt.minimize(
+            min_func,
+            x0=x0,
+            args=args,
+            jac=jac,
+            method=method,
+            options=options,
+        )  # Returns OptimizeResult
+        res_dict.update({"result": min_res})  # Updates res dict with side effect
+        return min_res.x.astype(x0.dtype)  # Return for host_callback
+
+    # hcb call with side effects to get the OptimizeResult on the same device it was called
+    hcb.call(
+        fun,
+        arg=None,
+        result_shape=x0,
+        call_with_device=isinstance(x0, jax.interpreters.xla.DeviceArray),
```

Comment:
What is this doing? "`call_with_device` – if True then the callback function is invoked with the device from which the call originates as a keyword argument." But `fun` doesn't do anything with its arguments, right?

Reply:
Actually, I have to look into this again due to the change I did to the device input of fun. What I want to achieve with this is that when I call the hcb function, even though it doesn't actually do anything with the argument, the execution of the hcb returns the output to the same device it was called from and not to the default device.

Reply:
I believe I fixed it, I will push what I have right now and test it from an environment with more than one device to ensure it's working properly. I will add some tests that reflect the device effect as well.

Reply:
Added.